### PR TITLE
Make templates return results properly instead of abusing .Scratch

### LINKED ISSUE #137: markdownify.html should not return result via scratch
It's rather awkward and inconvenient that `markdownify.html` puts its result in `.Scratch`, it should use `return` instead.

### PR DESCRIPTION
**Pull request by palant**
The initial commit fixes #137. I also could fail to notice that `hrefTargetBlank` processing in `markdownify.html` seems very inefficient - let me know if I missed something that the original code was doing better.

I must say that I didn't test the changes to `content.html` fully - it's way too many branches and settings. But the changes are pretty straightforward.

I want to add more commits to this pull request, there are other templates following the same pattern.

### CODE REVIEW

**Comment by palant** [comment]
Somewhat more extensive changes to `auto-detect-images.html` and `images.html` - these implemented pretty much the same logic concerning the image host. Now this logic is applied to the entire list of images, regardless of where the images came from. Also, `auto-detect-images.html` template now only executes as a fallback instead of executing even when it isn't needed.

**Comment by reuixiy** [comment]
> I also could fail to notice that `hrefTargetBlank` processing in `markdownify.html` seems very inefficient

Yeah... That was awkward, I'm not sure why I wrote that way now. 🤣

> I must say that I didn't test the changes to `content.html` fully

I have tested it and everything works fine.

> Somewhat more extensive changes to `auto-detect-images.html` and `images.html`

Love this refactoring!

**Comment by palant** [comment]
Side-notes: `title.html` will only pass page titles through `plainify`, not other titles - looks wrong. Also, `images.html` should probably recognize image names from the page bundle, something for another pull request.

**Comment by palant** [comment]
The changes to `relative-url.html` are more extensive again. The `./` prefix is unnecessary for a relative URL, so the template could be simplified. I also got rid of an unnecessary `safeJS` call in `service-worker.html` - I could verify that the code was emitted correctly here but I cannot test the functionality.

**Comment by palant** [comment]
It seems that I am done here. The change to `utils/tree-sections.html` is completely untested however because I have no idea under which circumstances this template is being used.

**Comment by reuixiy** [comment]
[`categoryBy = "sections"`](https://github.com/reuixiy/hugo-theme-meme/blob/9f506aef0ec125cbfd67d139b269129ee988b429/config-examples/en/config.toml#L382)
↳ [`layouts/section/categories.html`](https://github.com/reuixiy/hugo-theme-meme/blob/master/layouts/section/categories.html)
↳ [`layouts/partials/pages/tree-sections.html`](https://github.com/reuixiy/hugo-theme-meme/blob/master/layouts/partials/pages/tree-sections.html)
↳ Generated something like [this page](https://themes.gohugo.io/theme/hugo-theme-meme/categories/) but for [sections](https://gohugo.io/content-management/sections/).

In other words, I want to implement a categories page that displays like the output of `tree` command. So I created `utils/tree-sections.html` for this, although I don't know if there is a better way.

**Comment by reuixiy** [comment]
@palant Everything works well except the last commit, which returns an empty `map[]`.

Also, there are more `.Scratch`s we can remove, something for another pull request again. (To be honest, I didn't know the `=` operator until I found it in your commits.)

BTW, do you mind if I invite you as a collaborator of this repository? 😊

**Comment by palant** [comment]
Ok, I rolled back that last change. It's probably that one scenario where one actually has to use `.Scratch` - blocks cannot access external variables, and they cannot return a result either.

>  ↳ Generated something like this page but for sections.

Yes, I got this in theory but couldn't make that page generated.

> BTW, do you mind if I invite you as a collaborator of this repository? :blush: 

I guess that's ok but I cannot promise regular contributions.

**Comment by reuixiy** [comment]
> I guess that's ok but I cannot promise regular contributions.

No problem. I have invited you, and after you accept it, you should be able to merge this PR. 😊

Additionally, the commit messages, please follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/), as MemE's changelog is based on it.

Thanks again :)

**Comment by palant** [comment]
So the title of this commit should be "BREAKING CHANGE: Make templates return results properly instead of abusing .Scratch (#153)", right?

**Comment by reuixiy** [comment]
Hmmm, `refactor` sounds better,
```
refactor: Make templates return results properly (#153)

...instead of abusing .Scratch

Fixes #137

[...]
```

**Comment by palant** [comment]
Well, it is a breaking change for anybody overriding templates :smile:

**Comment by reuixiy** [comment]
Then maybe we can add it in the footer, I guess.

```
refactor: Make templates return results properly (#153)

...instead of abusing .Scratch

Fixes #137

[...]

BREAKING CHANGE: <description>
```
